Added support for QMX kernels in MLAS

qti-vaiskv opened the pull request:
Supported operations with QMX: SGEMM, QGEMM, Convolution

qti-vaiskv commented on onnxruntime/core/mlas/lib/qgemm.cpp:
```diff
@@ -205,8 +205,14 @@ bool
-#if defined(USE_KLEIDIAI) && !defined(_MSC_VER)
-  return ArmKleidiAI::UseSME2;
+#if defined(USE_KLEIDIAI)
+  if(ArmKleidiAI::UseSME2) {
```
Updated

qti-vaiskv commented on cmake/external/onnxruntime_external_deps.cmake:
```diff
@@ -845,6 +845,12 @@ if(onnxruntime_USE_KLEIDIAI)
   onnxruntime_fetchcontent_declare(kleidiai URL ${DEP_URL_kleidiai} URL_HASH SHA1=${DEP_SHA1_kleidiai} EXCLUDE_FROM_ALL)
   onnxruntime_fetchcontent_makeavailable(kleidiai)
+  # Fetch Qualcomm's kleidiai library
+  if(ENABLE_KLEIDIAI_QMX_COEXIST)
```
Currently, we are compiling by passing `--cmake_extra_defines MLAS_ENABLE_KLEIDIAI_QMX_COEXIST=ON` in the build command.

qti-vaiskv replied:
Added the option in build.py (`--use_qmx`)

qti-vaiskv commented on onnxruntime/core/mlas/lib/kleidiai/convolve_kleidiai.cpp:
```diff
@@ -597,10 +601,29 @@ static void ConvolveSme(const size_t co, //channels out
+                                    -std::numeric_limits<float>::max(), std::numeric_limits<float>::max()
+                                );
+                    }
+#else
```
Actually, here we are  conditionally compiling QMX & SME1 kernels if compiled with `MLAS_ENABLE_KLEIDIAI_QMX_COEXIST=ON` (which is the #if part). If this compile time option is not set, only SME1 kernel is compiled (which is in #else part)

qti-vaiskv commented on onnxruntime/core/mlas/lib/kleidiai/convolve_kleidiai.cpp:
```diff
@@ -597,10 +601,29 @@ static void ConvolveSme(const size_t co, //channels out
-                    TileSizeM, TileSizeN, d_kh * d_kw, ci, ATile, BTile, CTile, co * sizeof(float),
-                    -std::numeric_limits<float>::max(), std::numeric_limits<float>::max()
-                );
+#if(ENABLE_QMX_KERNELS)
```
Done

qti-vaiskv commented on onnxruntime/core/mlas/lib/kleidiai/qgemm_kleidiai.cpp:
```diff
@@ -108,14 +112,53 @@ ArmKleidiAI::MlasDynamicQGemmBatch(
+                    }
+                #else
+                        KLEIDIAI_KERNEL_LOG("kai_run_matmul_clamp_f32_qai8dxp1vlx4_qsi8cxp4vlx4_1vlx4vl_sme_mopa");
+                                kai_run_matmul_clamp_f32_qai8dxp1vlx4_qsi8cxp4vlx4_1vlx4vl_sme_mopa(
```
Done

qti-vaiskv commented on tools/ci_build/build_args.py:
```diff
@@ -763,6 +763,12 @@ def add_execution_provider_args(parser: argparse.ArgumentParser) -> None:
     )
 
+    # --- Qualcomm QMX Library ---
+    qmx_group = parser.add_argument_group("QMX kernel library")
```
Actually I have added this option in the QMX fork of Kleidiai library's CMakeLists.txt file

https://github.com/qualcomm/kleidiai/commit/85927d96456a18b9fd3b7d0f17bc2f3e0163f429

<img width="1304" height="165" alt="image" src="https://github.com/user-attachments/assets/b106068b-41a0-494c-8177-6a094331e2ee" />

I think similar to other Kleidiai config option, it should be part of QMX library's Cmake as the same option can be re-used in other runtime frameworks

hariharans29 replied:
Is the ENABLE_KLEIDIAI_QMX_COEXIST option for the QMX repo that is being passed through ?

Can we have an ORT CMake option to go with the build.py option taht if set turns on the appropriate flag in the QMX dependency ? I **think** that keeps it consistent with how things are currently in the ORT repo ?

hariharans29 replied:
Sorry not a cmake expert - but can the QMX cmake option -DENABLE_KLEIDIAI_QMX_COEXIST=ON be turned on conditionally within cmake for the QMX project when onnxruntime_USE_QMX_KLEIDIAI_COEXIST=ON is true ?

qti-vaiskv replied:
Yes, that is possible. I made the changes to set `ENABLE_KLEIDIAI_QMX_COEXIST` from QMX library based on Onnxruntime config option `onnxruntime_USE_QMX_KLEIDIAI_COEXIST`

hariharans29 commented:
Overall LGTM - thanks. There are some suggestions by Copilot and there is one minor formating comment by me. Can you please see if they are relevant ? Thanks.

qti-vaiskv commented on onnxruntime/core/mlas/lib/kai_ukernel_interface.cpp:
```diff
@@ -142,7 +160,17 @@ const kai_matmul_clamp_f32_f32p_f32p_ukernel& GetKleidiAISGemmUKernel() {
+        }
+#else
         return sgemm_gemm_sme;
+#endif // ENABLE_QMX_KERNELS
```
`GetKleidiAISGemmUKernel()` will be called only once. So string comparison will not be performance bottle neck.

qti-vaiskv commented:
> Overall LGTM - thanks. There are some suggestions by Copilot and there is one minor formating comment by me. Can you please see if they are relevant ? Thanks.

Hi @hariharans29 - Thanks for helping with the review. I have addressed the code indentation issues and few of the copilot comments as well.  The unaddressed comments from copilot, I don't think are very much relevant :)

hariharans29 commented:
/azp run Linux QNN CI Pipeline,Win_TRT_Minimal_CUDA_Test_CI,Windows ARM64 QNN CI Pipeline,Windows GPU Doc Gen CI Pipeline

hariharans29 commented:
Hello @qti-vaiskv - Can you please resolve the conflicts ? I ll merge this PR next. Thanks.

hariharans29 commented on onnxruntime/core/mlas/lib/kleidiai/mlasi_kleidiai.h:
```diff
@@ -54,6 +54,10 @@ namespace ArmKleidiAI {
+<<<<<<< HEAD
+inline const std::string_view vendor_name = MLAS_CPUIDINFO::GetCPUIDInfo().GetCPUVendor();
+=======
+>>>>>>> ms/main
```
Ditto

hariharans29 commented on cmake/deps.txt:
```diff
@@ -56,5 +56,8 @@ extensions;https://github.com/microsoft/onnxruntime-extensions/archive/c24b7bab0
 cudnn_frontend;https://github.com/NVIDIA/cudnn-frontend/archive/refs/tags/v1.12.0.zip;7e733cfdc410d777b76122d64232499205589a96
 dawn;https://github.com/google/dawn/archive/13c1635a14574ebb7116b56a69f5519301417fda.zip;0aadd28fc385cf7d657d5fc70a352372d2d3c76a
-kleidiai;https://github.com/ARM-software/kleidiai/archive/refs/tags/v1.20.0.tar.gz;6895e72b3d5cf1173358164cb3d64c9d7d33cc84
+kleidiai;https://github.com/ARM-software/kleidiai/archive/refs/tags/v1.15.0.tar.gz;62ccd24ab60bcef68766440fb42d79071ac2a5d2
```
Can we keep KAI version at 1.20.0 or can QMX only co-exist only with 1.15.0 ?
